### PR TITLE
[LIVY-620] Spark batch session always ends with success when configuration is master yarn and deploy-mode client

### PR DESCRIPTION
**Pull request by gumartinm**
## What changes were proposed in this pull request?

Batch session should end with state dead when process exits with no 0 return code.
https://issues.apache.org/jira/browse/LIVY-620

## How was this patch tested?

1. Unit Test (included in this PR)
Submit batch session that runs forever, wait 2 seconds, kill that batch session and expect for state dead.

2. Also currently used in production environment.

### CODE REVIEW

**Comment by jerryshao** [comment]
@yiheng please help to review.

**Comment by jerryshao** [comment]
Because we only get state from RM, so it always return SUCCESS in yarn client mode, this is the same as directly using spark-submit. Usually we suggest user to use yarn cluster mode.

File: server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala [hunk 1]
```diff
@@ -182,6 +183,14 @@ class BatchSession(
   override def stateChanged(oldState: SparkApp.State, newState: SparkApp.State): Unit = {
     synchronized {
       debug(s"$this state changed from $oldState to $newState")
+      if (_state != SessionState.Dead()) {
```
Comment (yiheng) [inline server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala]: Should we also don't allow state change when the state is other finished states? Maybe
```scala
if (!_state.isInstanceOf[FinishedSessionState]) {
    ...
}
```

File: server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala [hunk 2]
```diff
@@ -101,6 +101,7 @@ object BatchSession extends Logging {
             case 0 =>
             case exitCode =>
               warn(s"spark-submit exited with code $exitCode")
+              s.stateChanged(SparkApp.State.KILLED)
```
Comment (yiheng) [inline server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala]: why not use `SparkApp.State.FAILED`?

File: server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala (hunk 1 above)
Comment (gumartinm) [reply server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala]: You are absolutely right. I made the requested change. Thanks a lot for taking the time of reviewing this PR :)

File: server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala (hunk 2 above)
Comment (gumartinm) [reply server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala]: I had to choose either FAILED or KILLED. No idea of what you guys would like best. I've just changed it to FAILED.

Thanks again!!!

**Comment by gumartinm** [comment]
Retriggering build because Travis is failing with:

Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not transfer artifact some_artifact-SNAPSHOT from/to apache-snapshots (https://repository.apache.org/snapshots/): Connect to repository.apache.org:443 [repository.apache.org/207.244.88.140] failed: Connection timed out (Connection timed out)

**Comment by gumartinm** [comment]
Retriggering build because Travis is failing with:

[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.2:report-aggregate (jacoco-report) on project livy-coverage-report: An error has occurred in JaCoCo Aggregate report generation. Error while creating report: null: EOFException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.2:report-aggregate (jacoco-report) on project livy-coverage-report: An error has occurred in JaCoCo Aggregate report generation.


:/

**Comment by jerryshao** [comment]
There's some problems with repo we used recently, will fix that issue soon.

**Comment by jerryshao** [comment]
Retrigger the test.

**Comment by jerryshao** [comment]
LGTM, merging to master branch, thanks for the contribution.

**Pull request closed without merging**